Default blpAuthenticate object support redux

randomee opened the pull request:
cleanup of #283

Add support for default authentication object (modeled after support for default connection obj)

    - one new function - defaultAuthentication()
      more or less "sed s/Connection/Authentication/g < defaultConnection.R > defaultAuthentication.R"
    - export defaultAuthentication() to the NAMESPACE
    - change default args for "data" functions to use defaultAuthentication() not NULL
    - setup support for "blpAutoAuthenticate" akin to "blpAutoConnect".  Again, more or less copy/paste
    - documentation for default auth in the intro vignette

randomee commented:
Addresses @johnlaing 's note about saving a default auth object they way Rbbg did/does. #251

johnlaing commented:
My bad for dropping the ball on this a few months back, but thank you for following up. Here are some first-pass comments:

* I wasn't aware that SAPI with application ID was an option. Is that documented somewhere?

* We need to sort out the IP address resolution from host. The test `is.null(ip.address)` will now always be false because a default value is provided. Yet the default value of `localhost` will not work, as this is not an IP address.

* The `blpAuthenticate` function must continue to return its identity, even in the `default=TRUE` case. Existing code relied on getting an identity back without having to specify `default=FALSE`; we can't break that. We could return it invisibly, however.

* The init workflow needs to be tweaked. What happens if we try to auto-authenticate without having already connected? I'm sure this will fail, but I'm not quite sure where and whether it will be easy to diagnose.

randomee commented:
No worries.  We're all busy.  And I'm just scratching my own itch. 

I never looked for application-id auth over SAPI in the docs.  BBG support said to use it, and, well, it seems to work.  (Thank you @alfredkanzler for getting it working before I knew it was an option or needed!)

IP address, I'll look at.  

I think blpAuthenticate is behaving in a way that will continue to work.  Is this what you mean?
```R
> testIP <- "x.x.x.x"
> testUUID <- "xxxxxxxx"
> testSAPI <- "y.y.y.y"
>
> library(Rblpapi, quietly=TRUE)
Rblpapi version 0.3.10 using Blpapi headers 3.8.18.1 and run-time 3.8.18.1.
Please respect the Bloomberg licensing agreement and terms of service.
> blpConnect(host=testSAPI)
> blpid<-blpAuthenticate(ip=testIP, uuid=testUUID)
> bdp("IBM US Equity", "NAME", identity=blpid)
                                     NAME
IBM US Equity INTL BUSINESS MACHINES CORP
> bdp("IBM US Equity", "NAME")
                                     NAME
IBM US Equity INTL BUSINESS MACHINES CORP
> blpid
<pointer: 0x55f1c411cbf0>
>
```

The error for no-blpAutoConnect but with blpAutoAuthenticate seems like it has enough info to point users the right way.  Is this enough, or should it be more explicit?
```R
> testIP <- "x.x.x.x"
> testUUID <- "xxxxxxxx"
> testSAPI <- "y.y.y.y"
> options("uuid" = testUUID)
> options("blpIP" = testIP)
> options("blpHost" = testSAPI)
> options("blpAutoAuthenticate" = TRUE)
> library(Rblpapi, quietly=TRUE)
Rblpapi version 0.3.10 using Blpapi headers 3.8.18.1 and run-time 3.8.18.1.
Please respect the Bloomberg licensing agreement and terms of service.
Error: package or namespace load failed for ‘Rblpapi’:
 .onAttach failed in attachNamespace() for 'Rblpapi', details:
  call: NULL
  error: No connection object has been created. Use 'blpConnect()' first.
>
```

randomee commented:
Would changing the localhost code from is.null() to a compare be reasonable?  

```R
if (is.null(ip.address)) {
```

changed to
```R
if (ip.address == "localhost") {
```

None of the setups I have access to run SAPI on localhost, so I can't really test it fully.  I can make the commit for someone else to verify though.

randomee commented:
Looking at the localhost bit, I don't quite understand why we're changing "host" when "ip.address" is localhost.  
```R
        if (missing(ip.address)) {
            ## Linux only ?
            cmd.res <- system(paste("host",host), intern=TRUE,
                              ignore.stdout=FALSE, ignore.stderr=FALSE,wait=TRUE)
            ip.address <- strsplit(cmd.res,"address ")[[1]][2]
}
```
As I understand things, "ip.address" is the last IP the uuid successfully authenticated from (e.g. the IP of their bloomberg terminal.  And "host" is the SAPI (or bpipe) server address. 

It seems like this is a shortcut to say the SAPI server is localhost if you don't provide where you last authenticated.  That strikes me as non-intuitive. Do you know if that's how it's being used?  

I will fully admit I've not read the SAPI/API docs in depth and could be totally missing something obvious.

johnlaing commented:
You are right - it does still work. I had anticipated that it might not return the connection object if `default=FALSE` but it does. I think we could make that a little more clear, however.

I think you may be confusing the `host` argument between `blpConnect` and `blpAuthenticate`. In the former it refers to the SAPI server, but in the latter it refers to whatever machine you are logged in from. In fact, the authentication service wants an IP address which can be passed explicitly via `ip.address`. We also provide the `host` argument as a convenience. (My workstation name never changes, but its local IP might from time to time.)

I think the test `is.null(ip.address)` is good, we just have to change the default argument to `ip.address=getOption("blpIP", NULL)`

randomee commented:
Ah, ok.  I never groked that the hostname here is not the same as for blpConnect.  I don't think the documentation on that is clear at all.  In that case, perhaps there should be an error/warning thrown if both IP and host are passed in.  Silently taking the IP ,when the hostname may resolve differently, seems like a potentially really annoying problem to debug. 

That also means `ip.address=getOption("blpIP", NULL)` is incorrect.  And there should be a separate default `options()` variable.  Maybe blpServerHostname?  kinda verbose, but clearer.

Changed, and pushed the NULL default.  I have not had a chance to test it, yet.  I'll update when I have.  It still needs documentation and code changes for the above discussion points.

eddelbuettel commented:
This seems to move nicely in the right direction as everybody can still run it.  @randomee an entry in file [ChangeLog](https://github.com/Rblp/Rblpapi/blob/master/ChangeLog) would be appreciated too.  (Uses an old GNU/FSF format that Emacs happens to make easy; two spaces around name and email. Of course you can mock it by hand in another editor ....)

randomee commented:
- Cleaned up documentation
- added warning if both ip.address and host are passed to `blpAuthenticate()`
- renamed the `options()` variable names to make the host vs host distinction clearer
- renamed `options(uuid)` to `blpUUID` to keep the namespace more consistient

I'll test this tomorrow on a variety of machines and confirm that it is working where I can test it.

randomee commented:
Looks reasonable and working for a good sample set. All SAPI & DAPI calls are against one company's infra.  I'll test 2 more tomorrow.  If there are errors, I'll comment. 

If there are more code paths to test, let me know and I'll see what I can do.

I cannot test against B-PIPE, or SAPI running on localhost.  If someone could check those, that would be nice.  

Otherwise, I think it's in pretty good shape.  It would be nice to run `roxygen2::roxygenise()` to update the various docs.  I can do that, or leave it for the next release cycle process.

With SAPI, uuid, manual connect, & auth
Note warning for supplying both IP & hostname.
```R
> testIP <- "x.x.x.x"
> testUUID <- "yyyyyyy"
> testSAPI <- "z.z.z.z"
> options("blpUUID" = testUUID)
> options("blpLoginIP" = testIP)
> options("blpHost" = testSAPI)
> options("blpPort" = 8194)
> options("blpLoginHostname" = testIP)
> library(Rblpapi, quietly=TRUE)
Rblpapi version 0.3.10 using Blpapi headers 3.8.18.1 and run-time 3.8.18.1.
Please respect the Bloomberg licensing agreement and terms of service.
> blpConnect()
> blpAuthenticate()
Warning message:
In blpAuthenticate() : Both ip.address and host are set.  Using ip.address.
> bdp("IBM US Equity", "NAME")
                                     NAME
IBM US Equity INTL BUSINESS MACHINES CORP
>
```

With SAPI, uuid, auto connect & auth
```R
> testIP <- "x.x.x.x"
> testUUID <- "yyyyyyyy"
> testSAPI <- "z.z.z.z"
>
> options("blpUUID" = testUUID)
> options("blpLoginIP" = testIP)
> options("blpHost" = testSAPI)
> options("blpPort" = 8194)
>
> options("blpAutoConnect" = TRUE)
> options("blpAutoAuthenticate" = TRUE)
> library(Rblpapi, quietly=TRUE)
Rblpapi version 0.3.10 using Blpapi headers 3.8.18.1 and run-time 3.8.18.1.
Please respect the Bloomberg licensing agreement and terms of service.
> bdp("IBM US Equity", "NAME")
                                     NAME
IBM US Equity INTL BUSINESS MACHINES CORP
>
```

SAPI, uuid, manual connect & auth
```R
> testIP <- "x.x.x.x"
> testUUID <- "yyyyyyyy"
> testSAPI <- "z.z.z.z"
> library(Rblpapi, quietly=TRUE)
Rblpapi version 0.3.10 using Blpapi headers 3.8.18.1 and run-time 3.8.18.1.
Please respect the Bloomberg licensing agreement and terms of service.
> blpConnect(host=testSAPI)
> blpAuthenticate(uuid=testUUID,ip.address=testIP)
> bdp("IBM US Equity", "NAME")
                                     NAME
IBM US Equity INTL BUSINESS MACHINES CORP
>
```

With DAPI, auto connect
```R
> testDesktop<-"testdesktop"
> options("blpHost" = testDesktop)
> options("blpAutoConnect" = TRUE)
> library(Rblpapi, quietly=TRUE)
Rblpapi version 0.3.10 using Blpapi headers 3.8.18.1 and run-time 3.8.18.1.
Please respect the Bloomberg licensing agreement and terms of service.
> bdp("IBM US Equity", "NAME")
                                     NAME
IBM US Equity INTL BUSINESS MACHINES CORP
>
```

With DAPI, manual connect
```R
> library(Rblpapi, quietly=TRUE)
Rblpapi version 0.3.10 using Blpapi headers 3.8.18.1 and run-time 3.8.18.1.
Please respect the Bloomberg licensing agreement and terms of service.
> blpConnect(host="testdesktop")
> bdp("IBM US Equity", "NAME")
                                     NAME
IBM US Equity INTL BUSINESS MACHINES CORP
>

```

SAPI, with Application-name, autoconnect & auth
```R
> options("blpHost" = "x.x.x.x")
> options("blpAppName" = "Foo:Bar")
> options("blpAutoConnect" = TRUE)
> options("blpAutoAuthenticate" = TRUE)
>
> library(Rblpapi, quietly=TRUE)
Rblpapi version 0.3.10 using Blpapi headers 3.8.18.1 and run-time 3.8.18.1.
Please respect the Bloomberg licensing agreement and terms of service.
> bdp("IBM US Equity", "NAME")
                                     NAME
IBM US Equity INTL BUSINESS MACHINES CORP
>

```

SAPI, Application name, manual connect
```R
> options("blpHost" = "x.x.x.x")
> options("blpAppName" = "Foo:Bar")
> library(Rblpapi, quietly=TRUE)
Rblpapi version 0.3.10 using Blpapi headers 3.8.18.1 and run-time 3.8.18.1.
Please respect the Bloomberg licensing agreement and terms of service.
> blpConnect()
> blpAuthenticate()
> bdp("IBM US Equity", "NAME")
                                     NAME
IBM US Equity INTL BUSINESS MACHINES CORP
>
```

eddelbuettel commented on vignettes/rblpapi-intro.Rmd:
```diff
@@ -48,18 +48,79 @@ blpConnect()
+options("blpAutoConnect" = TRUE)
+options("blpHost" = "x.x.x.x")
+options("blpPort" = 8194)
+options("blpAppName" = "yyy")
```
Should we say that AppName is optional and used only for B-Pipe?   (If I have that right?)

randomee replied:
edited.  have a look.

eddelbuettel replied:
Thanks!

eddelbuettel commented on vignettes/rblpapi-intro.Rmd:
````diff
@@ -48,18 +48,79 @@ blpConnect()
+```
+
+Usually needed for SAPI & B-PIPE sessions.  Not usually needed for DAPI
+(Desktop) sessions.
````
Could we make this sentence, and the following ones, complete sentences please?

randomee replied:
edited.  have a look.

eddelbuettel replied:
Thanks!

(Technically, I think, the person raising the issue tends to call it as resolved.  No worries though, appreciate the quick reply.)

randomee replied:
mea culpa

eddelbuettel commented on ChangeLog:
```diff
@@ -1,3 +1,23 @@
+2019-05-23  randomee  <randomee@users.noreply.github.com>
```
We would really prefer a real name and a resolvable email address here.

randomee replied:
Honestly, I'd prefer not to put that info in public.  Past bad on-line to off-line issues.  I'm happy to give you info backchannel, or meet in the loop next time I'm downtown.

eddelbuettel replied:
While I (nominally) understand where you are coming from, allow me to state that we are the (very public) face of this code, and have been for a decade (or longer if you count predecessor).  People will come to John, Whit, myself or now Alfred because there is an identifiable counterparty as well as some trust and visibility.   

I simply have a harder time standing behind code from a random unidentified strangers.  So as (current) Maintainer, I may be inclined to say no to this pull request for this reason.   At the end of the day I simply have no idea who you are, or where you are from.  We will discuss, but it would also be good if you could give this some thought over the long weekend, or if need be, the next few weeks.

eddelbuettel commented:
Just quick heads-up to say that this has not fallen off the wagon ... but time is tight, markets are buys and all that.  We'll get to it, hopefully 'soon'.

randomee commented:
No worries.  I'll likely be largely MIA in August, just FYI.

johnlaing commented:
I refactored some of the interaction between connection/authentication, updated docs, etc. This now seems to be working on my end but it would be good for others to test as well.

randomee commented:
Sorry, busy, etc.  

The code looks correct, but obviously should be tested.  I'll try to test this weekend.  I can test them on:

- DAPI
- SAPI - no auth - linux
- SAPI - uuid/ip auth - linux
- SAPI - appid auth - linux

Hopefully that's enough coverage.

Thanks for bumping this.

randomee commented:
Ok.  Looks good.  Tested various ways:

- auto-connecting
- manual connection
- passing id object (for SAPI/uuid and SAPI/appid)
- using default id object (for SAPI/uuid and SAPI/appid)

Tested these BBG backend types:
- DAPI - **OK**
- SAPI - no auth - linux - **OK**
- SAPI - appid auth - linux - **OK**
- SAPI - uuid/ip auth - linux - **OK**
